### PR TITLE
Rename combinators, introduce CompositeLengthException

### PR DESCRIPTION
**Pull request by kelunik**

### CODE REVIEW

File: src/Future/functions.php [hunk 1]
```diff
@@ -29,39 +29,82 @@ function race(iterable $futures, ?Cancellation $cancellation = null): mixed
...
  * @param iterable<Tk, Future<Tv>> $futures
- * @param Cancellation|null        $cancellation Optional cancellation.
+ * @param Cancellation|null $cancellation Optional cancellation.
  *
  * @return non-empty-array<Tk, Tv>
  *
  * @throws CompositeException If all futures errored.
  */
-function some(iterable $futures, int $count, ?Cancellation $cancellation = null): array
+function awaitN(iterable $futures, int $count, ?Cancellation $cancellation = null): array
```
Comment (kelunik) [inline src/Future/functions.php:107]: Should we switch the parameter order here, so `$count` will be the first param?

File: src/Future/functions.php [hunk 2]
```diff
@@ -81,7 +124,7 @@ function some(iterable $futures, int $count, ?Cancellation $cancellation = null)
         }
     }
 
-    if (empty($errors)) {
+    if (\count($values) + \count($errors) < $count) {
```
Comment (kelunik) [inline src/Future/functions.php:127]: I'm not sure whether this should really be an `\Error`, IMO this should probably be an exception, as the consumer might not know how many elements are in the iterator.

Comment (bwoebi) [reply src/Future/functions.php:127]: I agree - an Exception is better.

File: src/Future/functions.php (hunk 1 above)
Comment (trowski) [reply src/Future/functions.php:107]: Agreed.

Comment (kelunik) [reply src/Future/functions.php:107]: Should this be named `awaitAnyN`?

Comment (bwoebi) [reply src/Future/functions.php:107]: possibly there should be awaitN (= after N futures completed, we're done) and awaitAnyN (= after any N futures succeeded, we're done)?

Comment (kelunik) [reply src/Future/functions.php:107]: Yeah, discussed this with @trowski privately as well, we could introduce `awaitFirstN` later if needed.

File: src/Future/functions.php (hunk 2 above)
Comment (kelunik) [reply src/Future/functions.php:127]: Which exception should we throw here? `CompositeException` currently requires a non-empty list of exceptions, but I'm not sure whether that's a good idea, as we could use it here without having any errors if there aren't any futures in the iterable.

Comment (trowski) [reply src/Future/functions.php:127]: Seems fine to allow an empty list.

**Review by trowski — APPROVED** [review]

**Pull request merged**